File: IPAddress/src/inet/ipaddr/format/validate/MACAddressProvider.java
/*
 * Copyright 2017 Sean C Foley
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *     or at
 *     https://github.com/seancfoley/IPAddress/blob/master/LICENSE
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package inet.ipaddr.format.validate;

import java.io.Serializable;

import inet.ipaddr.mac.MACAddress;
import inet.ipaddr.mac.MACAddressSegment;

public class MACAddressProvider implements Serializable {

	private static final long serialVersionUID = 3L;
	
	private ParsedMACAddress parsedAddress;
	private MACAddress address;
	
	static final MACAddressProvider EMPTY_PROVIDER = new MACAddressProvider() {
		
		private static final long serialVersionUID = 3L;

		@Override
		public MACAddress getAddress() {
			return null;
		}
	};
	
	private static final MACAddress ALL_MAC_ADDRESSES = new MACAddress(new MACAddressSegment[] {
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT
	});
	
	static final MACAddressProvider ALL_MAC = new MACAddressProvider() {

		private static final long serialVersionUID = 3L;

		@Override
		public MACAddress getAddress() {
			return ALL_MAC_ADDRESSES;
		}
	};
	
	private static final MACAddress ALL_EUI_64_ADDRESSES = new MACAddress(new MACAddressSegment[] {
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT,
		MACAddressSegment.ALL_RANGE_SEGMENT
	});
	
	static final MACAddressProvider ALL_EUI_64 = new MACAddressProvider() {

		private static final long serialVersionUID = 3L;

		@Override
		public MACAddress getAddress() {
			return ALL_EUI_64_ADDRESSES;
		}
	};
	
	private MACAddressProvider() {}
	
	public MACAddressProvider(ParsedMACAddress parsedAddress) {
		this.parsedAddress = parsedAddress;
	}
	
	public MACAddressProvider(MACAddress address) {
		this.address = address;
	}

	public MACAddress getAddress() {
		if(parsedAddress != null) {
			synchronized(this) {
				if(parsedAddress != null) {
					address = parsedAddress.createAddress();
					parsedAddress = null;
				}
			}
		}
		return address;
	}
	
	@Override
	public String toString() {
		return String.valueOf(getAddress());
	}
}

File: ipaddress-go/ipaddr/macaddressprovider.go
//
// Copyright 2020-2021 Sean C Foley
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

package ipaddr

import (
	"sync"

	"github.com/seancfoley/ipaddress/ipaddress-go/ipaddr/addrerr"
	"github.com/seancfoley/ipaddress/ipaddress-go/ipaddr/addrstrparam"
)

type macAddressProvider interface {
	getAddress() (*MACAddress, addrerr.IncompatibleAddressError)
}

type macAddressEmptyProvider struct{}

func (provider macAddressEmptyProvider) getAddress() (*MACAddress, addrerr.IncompatibleAddressError) {
	return nil, nil
}

var defaultMACAddressEmptyProvider = macAddressEmptyProvider{}

type macAddressAllProvider struct {
	validationOptions addrparam.MACAddressStringParams
	address           *MACAddress
	creationLock      sync.Mutex
}

func (provider *macAddressAllProvider) getAddress() (*MACAddress, addrerr.IncompatibleAddressError) {
	addr := provider.address
	if addr == nil {
		provider.creationLock.Lock()
		addr = provider.address
		if addr == nil {
			validationOptions := provider.validationOptions
			//creator := provider.validationOptions.GetNetwork().GetMACAddressCreator() xxxx ipaddress we used addrType to get network to get creator
			size := validationOptions.GetPreferredLen()
			creator := macType.getNetwork().getAddressCreator()
			var segCount int
			if size == addrparam.EUI64Len {
				segCount = ExtendedUniqueIdentifier64SegmentCount
			} else {
				segCount = MediaAccessControlSegmentCount
			}
			allRangeSegment := creator.createRangeSegment(0, MACMaxValuePerSegment)
			segments := make([]*AddressDivision, segCount)
			for i := range segments {
				segments[i] = allRangeSegment
			}
			section := creator.createSectionInternal(segments, true)
			addr = creator.createAddressInternal(section.ToSectionBase(), nil).ToMAC()
		}
		provider.creationLock.Unlock()
	}
	return addr, nil
}

var macAddressDefaultAllProvider = &macAddressAllProvider{validationOptions: defaultMACAddrParameters}

type wrappedMACAddressProvider struct {
	address *MACAddress
}

func (provider wrappedMACAddressProvider) getAddress() (*MACAddress, addrerr.IncompatibleAddressError) {
	return provider.address, nil
}

var (
	_, _, _ macAddressProvider = &macAddressEmptyProvider{},
		&macAddressAllProvider{},
		&wrappedMACAddressProvider{}
)
